sql/moz-fx-data-shared-prod/udf/bits_from_offsets/udf.sql

```diff
@@ -0,0 +1,48 @@
+CREATE OR REPLACE FUNCTION udf.bits_from_offsets(offsets ARRAY<INT64>) AS (
+  (
+    SELECT
+      LTRIM(
+        STRING_AGG(
+          (
+            SELECT
+              -- CODE_POINTS_TO_BYTES is the best available interface for converting
+              -- an array of numeric values to a BYTES field; it requires that values
+              -- are valid extended ASCII characters, so we can only aggregate 8 bits
+              -- at a time, and then concat all those chunks together with STRING_AGG.
+              CODE_POINTS_TO_BYTES(
+                [
+                  BIT_OR(
+                    (
+                      IF(
+                        DIV(n - (8 * i), 8) = 0
+                        AND (n - (8 * i)) >= 0,
+                        1 << MOD(n - (8 * i), 8),
+                        0
+                      )
+                    )
+                  )
+                ]
+              )
+            FROM
+              UNNEST(offsets) AS n
+          ),
+          b''
+        ),
+        b'\x00'
+      )
+    FROM
+      -- Each iteration handles 8 bits, so 256 iterations gives us 2048 bits,
```

Comment:
An `activity_to_offset` function would need to also take in a `target_date` parameter.

So it would be the following:

```
  bits_from_offsets(
    ARRAY_AGG(udf.activity_to_offset(foo, anchor_date, submission_date) IGNORE NULLS)
  )
```

compared to the more raw:

```
  bits_from_offsets(
    ARRAY_AGG(IF(foo, DATE_DIFF(anchor_date, submission_date, DAY), NULL)
              IGNORE NULLS)
  )
```

Does the cleanliness of the `activity_to_offset` version warrant the extra indirection? I worry about higher possibility of getting the order of parameters wrong.

Reply:
We know the maximum `anchor_date` already: 2048 days ago. We should always be able to use that maximum, and let the `ltrim` deal with the unused portion. When/if we update this function to include 2048 days, we would also update the sibling function.

Not a requirement by any means, just an option.

Reply:
> We know the maximum anchor_date already: 2048 days ago

I still think this should be passed in. The anchor_date shouldn't be assumed based on the current date, but rather determined by the target date of execution of the query. If Airflow were to run a query from two days ago, for example, it would need to calculate bytes based on a time period ending on that previous execution date.